src/arena.rs

```diff
@@ -354,3 +355,414 @@ impl<T> Clone for List<T> {
+    }
+
+    /// Removes and returns the element at the front of this list.  If the list is empty, returns
+    /// `None`.
+    pub fn pop_front<'a>(&mut self, arena: &'a ReversibleListArena<T>) -> Option<&'a T> {
+        match arena.get(self.cells) {
+            ReversibleListCell::Empty => return None,
+            ReversibleListCell::Nonempty { head, tail, .. } => {
+                self.cells = *tail;
+                Some(head)
+            }
+        }
+    }
+
+    /// Returns an iterator over the elements of this list.
+    pub fn iter<'a>(&self, arena: &'a ReversibleListArena<T>) -> impl Iterator<Item = &'a T> + 'a {
+        let mut current = self.cells;
+        std::iter::from_fn(move || match arena.get(current) {
+            ReversibleListCell::Empty => return None,
+            ReversibleListCell::Nonempty { head, tail, .. } => {
+                current = *tail;
+                Some(head)
+            }
+        })
+    }
+}
+
+impl<T> ReversibleList<T>
+where
+    T: Clone,
+{
+    /// Reverses the list.  Since we're already caching everything in an arena, we make sure to
+    /// only calculate the reversal once, returning it as-is if you call this function multiple
+    /// times.
+    pub fn reverse(&mut self, arena: &mut ReversibleListArena<T>) {
+        self.ensure_reversal_available(arena);
+        match arena.get(self.cells) {
+            ReversibleListCell::Empty => return,
+            ReversibleListCell::Nonempty { reversed, .. } => match reversed.get() {
+                Some(reversed) => self.cells = reversed,
+                _ => unreachable!(),
+            },
+        }
+    }
+
+    /// Ensures that the reversal of this list is available.  It can be useful to precalculate this
+    /// when you have mutable access to the arena, so that you can then reverse and un-reverse the
+    /// list later when you only have shared access to it.
+    pub fn ensure_reversal_available(&mut self, arena: &mut ReversibleListArena<T>) {
+        // First check to see if the list has already been reversed.
+        match arena.get(self.cells) {
+            // The empty list is already reversed.
+            ReversibleListCell::Empty => return,
+            ReversibleListCell::Nonempty { reversed, .. } if reversed.get().is_some() => return,
+            _ => (),
+        }
+
+        // If not, reverse the list and cache the result.
+        let new_reversed = ReversibleListCell::reverse(self.cells, arena);
+        match arena.get(self.cells) {
+            ReversibleListCell::Nonempty { reversed, .. } => {
+                reversed.set(Some(new_reversed));
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<T> ReversibleList<T> {
+    /// Reverses the list, assuming that the reversal has already been computed.  If it hasn't we
+    /// return an error.
+    pub fn reverse_reused(&mut self, arena: &ReversibleListArena<T>) -> Result<(), ()> {
```

Comment:
I'd prefer to keep it as `()` because of YAGNI.  The main reason I'd see for using a more specific type is if we want to write `From` impls on _other_ error types that we could easily translate this error condition into.  (That would enable the `?` operator on the result of this method.)  Currently every time we call this function we're using `expect` on the result, so we don't need to do that.  In the future if we need to, we can add the more specific error type then.

Reply:
Sounds good to me. 👍🏻